lighthouse-core/scripts/i18n/collect-strings.js

```diff
@@ -30,18 +28,321 @@ const ignoredPathComponents = [
+    }
+  }
+}
+
+/**
+ * Convert markdown html links into placeholders.
+ *
+ * @param {ICUMessageDefn} icu
+ */
+function _processPlaceholderMarkdownLink(icu) {
+  // Check that link markdown is not slightly off.
+  if (icu.message.match(/\[.*\] \(.*\)/)) {
+    throw Error(`Bad Link syntax in message "${icu.message}"`);
+  }
+
+  // Split on markdown links (e.g. [some link](https://...)).
+  const parts = icu.message.split(/\[([^\]]*?)\]\((https?:\/\/.*?)\)/g);
+  icu.message = '';
+  let idx = 0;
+
+  while (parts.length) {
+    // Pop off the same number of elements as there are capture groups.
+    const [preambleText, linkText, linkHref] = parts.splice(0, 3);
+    icu.message += preambleText;
+
+    // Append link if there are any.
+    if (linkText && linkHref) {
+      const ls = `LINK_START_${idx}`;
+      const le = `LINK_END_${idx++}`;
+      const repr = `$${ls}$${linkText}$${le}$`;
+      icu.message += repr;
+      icu.placeholders[ls] = {
+        content: '[',
+      };
+      icu.placeholders[le] = {
+        content: `](${linkHref})`,
+      };
+    }
   }
+}
+
+/**
+ * Convert complex ICU syntax into placeholders with examples.
+ *
+ * @param {ICUMessageDefn} icu
+ */
+function _processPlaceholderComplexIcu(icu) {
+  // Check that complex ICU not using non-supported format
+  if (icu.message.match(
+    /\{(\w{2,50}), number, (?!milliseconds|seconds|bytes|percent|extendedPercent).*\}/)) {
+    throw Error(`Unsupported ICU format in message "${icu.message}"`);
+  }
+
+  // Split on complex ICU: {var, number, type}
+  const parts = icu.message.split(
+    /\{(\w{2,50}), number, (milliseconds|seconds|bytes|percent|extendedPercent)\}/g);
+  icu.message = '';
+  let idx = 0;
+
+  while (parts.length) {
+    // Pop off the same number of elements as there are capture groups.
+    const [preambleText, varName, icuType] = parts.splice(0, 3);
+    icu.message += preambleText;
+
+    // Append link if there are any.
+    if (varName && icuType) {
+      const iName = `COMPLEX_ICU_${idx++}`;
+      icu.message += `$${iName}$`;
+      let example = '0';
+
+      // Make some good examples.
+      switch (icuType) {
+        case 'seconds':
+          example = '2.4';
+          break;
+        case 'percent':
+          example = '54.6%';
+          break;
+        case 'extendedPercent':
+          example = '37.92%';
+          break;
+        default:
+          // Random (but constant) number for examples.
+          example = '499';
+      }
+
+      icu.placeholders[iName] = {
+        content: `{${varName}, number, ${icuType}}`,
+        example: example,
+      };
+    }
+  }
+}
+
+/**
+ * Add examples for direct ICU replacement.
+ *
+ * @param {ICUMessageDefn} icu
+ * @param {Record<string, string>} examples
+ */
+function _processPlaceholderDirectIcu(icu, examples) {
+  let tempMessage = icu.message;
+  let idx = 0;
+  for (const [key, value] of Object.entries(examples)) {
+    if (!icu.message.includes(`{${key}}`)) continue;
```

Comment:
So, this is a little jank.  Technically, when computing the description in `computeDescription()` it will throw if there aren't any examples.  `_processPlaceholderDirectIcu()` is dumb, and will just take any arbitrary list of examples and a string and calc the resultant string+placeholders.  It has no concept of required examples, that is done prior to it.  It keeps the direct ICU replacement function as short and as dumb as possible, and puts the heavy lifting onto the description calculator, which is already a heavy script that has a lot of rules and checking, etc.

Reply:
> Technically, when computing the description in `computeDescription()` it will throw if there aren't any examples. `_processPlaceholderDirectIcu()` is dumb, and will just take any arbitrary list of examples and a string and calc the resultant string+placeholders. It has no concept of required examples, that is done prior to it. It keeps the direct ICU replacement function as short and as dumb as possible, and puts the heavy lifting onto the description calculator, which is already a heavy script that has a lot of rules and checking, etc.

I think this is the right way to think about it, but I would split the concerns slightly differently. The issue with the current approach is that two places have to know how the different types of replacements work. This means, for instance, if we were to change the current requirements for direct replacement (an `@description` and `@example` for every placeholder in the string) we'd need to update `_processPlaceholderDirectIcu` (obvious) but also `computeDescription()` (less obvious where that would happen, unless you manage to notice the `findIcu` regex only matches direct replacement placeholders). Same thing if we wanted to do something like allow manual examples for formatted placeholders.

If, on the other hand, `computeDescription()` was only responsible for parsing the jsdoc comment (give it a comment, it passes back `description` and (if present) `examples`), that keeps that method more focused and independent of placeholder specifics, and meanwhile there's a natural place for checking the requirements for each kind of replacement that might be found since each has its own `_process*` function.

Reply:
Separated them out. `computeDesc` is much more dumb, will just give you whatever you have, and `_processDirect` has the checking logic in it! Updated tests as well.